asks/multipart.py

```diff
@@ -0,0 +1,146 @@
+
+    Returns:
+        MultipartData: An object describing a multipart/form-data file field.
+    '''
+
+    basename = Path(value.name).name if hasattr(value, 'name') else None
+    mime_type = (
+        mimetypes.guess_type(basename)[0]
+        if basename is not None
+        else _RAW_BYTES_MIMETYPE
+    )
+
+    return MultipartData(
+        binary_source=value,
+        mime_type=mime_type,
+        basename=basename,
+    )
+
+
+def _to_multipart_form_data(value, encoding):
+    '''
+    Transform a form-data entry into a MultipartData object.
+
+    If `value` is one of the supported file-like types, it will be seen as a
+    file.  Else it will be converted into a string and sent as raw form data.
+
+    Args:
+        value: A form value to encode or a file-like supported object.
+        encoding: The desired encoding to send a field as.
+
+    Returns:
+        MultipartData: An object describing a multipart/form-data field.
+    '''
+    if isinstance(value, MultipartData):
+        return value
+
+    if isinstance(value, (Path, bytes)) or hasattr(value, 'read'):
+        return _to_multipart_file(value)
+
+    # It's not a supported file type, so we do our best to transform it into form data.
+    return MultipartData(
+        binary_source=str(value).encode(encoding),
+        mime_type=None,
+        basename=None,
+    )
+
+
+async def build_multipart_body(values, encoding, boundary_data):
+    '''
+    Forms a multipart request body from  a dict of form fields to values.
+
+    Args:
+        values: A dict of strings (field names) to:
+                    - Path
+                    - IO[bytes]
+                    - An async file-like object (read() is async), opened in binary mode.
+                    - bytes
+                    - MultipartData
+                    - Any other value that can be sent by being converted to str.
+                The first four will be sent as files. MultipartData will be sent
+                as a file if it has the necessary info to do so. The rest will be
+                sent as ordinary data.
+
+    Returns:
+        multipart_body (bytes): The strings representation of the content body,
+                                multipart formatted.
+    '''
+    boundary = b''.join((b'--', bytes(boundary_data, encoding)))
+
+    multipart_body = b''
+
+    for k, v in values.items():
+        multipart_data = _to_multipart_form_data(v, encoding)
+
+        multipart_body += b''.join((
+            boundary,
+            b'\r\n',
+            'Content-Disposition: form-data; name="{}"'.format(k).encode(encoding),
+            (
+                b''
+                if multipart_data.basename is None
+                else '; filename="{}"'.format(multipart_data.basename).encode(encoding)
+            ),
+            (
+                b''
+                if multipart_data.mime_type is None
+                else '; Content-Type: {}'.format(multipart_data.mime_type).encode(encoding)
```

Comment:
This should probably be '\r\n Content-Type' instead of '; Content-Type' to actually create a new header for the body instead of something like this:
`Content-Disposition: form-data; name="file_field"; filename="example.txt"; Content-Type: text/plain^M`